src/device.rs

```diff
@@ -0,0 +1,229 @@
+pub struct DeviceList {
+    /// base address
+    arr: NonNull<Device>,
+    /// array length
+    len: usize,
+}
+
+/// SAFETY: owned array
+unsafe impl Send for DeviceList {}
+/// SAFETY: owned array
+unsafe impl Sync for DeviceList {}
+
+/// A RDMA device
+#[allow(missing_copy_implementations)] // This type can not copy
+#[repr(transparent)]
+pub struct Device(NonNull<ibv_device>);
+
+/// SAFETY: owned type
+unsafe impl Send for Device {}
+/// SAFETY: owned type
+unsafe impl Sync for Device {}
+
+/// A RDMA device guid
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Guid(__be64);
+
+impl DeviceList {
+    /// Returns `*mut *mut ibv_device`
+    fn ffi_ptr(&self) -> *mut *mut ibv_device {
+        self.arr.as_ptr().cast()
+    }
+
+    /// Returns available rdma devices
+    #[inline]
+    pub fn available() -> io::Result<Self> {
+        // SAFETY: ffi
+        unsafe {
+            let mut num_devices: c_int = 0;
+            let arr = ibv_get_device_list(&mut num_devices);
+            if arr.is_null() {
+                return Err(last_errno());
+            }
+
+            // SAFETY: repr(transparent)
+            let arr: NonNull<Device> = NonNull::new_unchecked(arr.cast());
+
+            let _guard = guard_on_unwind((), |()| ibv_free_device_list(arr.as_ptr().cast()));
+
+            let len: usize = num_devices.numeric_cast();
+
+            if mem::size_of::<c_int>() >= mem::size_of::<usize>() {
+                let total_size = len.saturating_mul(mem::size_of::<*mut ibv_device>());
+                assert!(total_size < usize::MAX.wrapping_div(2));
+            }
+
+            Ok(Self { arr, len })
+        }
+    }
+
+    /// Returns the slice of devices
+    #[inline]
+    #[must_use]
+    pub fn as_slice(&self) -> &[Device] {
+        // SAFETY: guaranteed by `DeviceList::available`
+        unsafe { slice::from_raw_parts(self.arr.as_ptr(), self.len) }
+    }
+}
+
+impl Drop for DeviceList {
+    #[inline]
+    fn drop(&mut self) {
+        // SAFETY: ffi
+        unsafe { ibv_free_device_list(self.ffi_ptr()) }
+    }
+}
+
+impl Deref for DeviceList {
+    type Target = [Device];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl fmt::Debug for DeviceList {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        <[Device] as fmt::Debug>::fmt(self, f)
+    }
+}
+
+impl Device {
+    /// Returns `*mut ibv_device`
+    pub(crate) fn ffi_ptr(&self) -> *mut ibv_device {
+        self.0.as_ptr()
+    }
+
+    /// Returns kernel device name
+    #[inline]
+    #[must_use]
+    pub fn c_name(&self) -> &CStr {
+        // SAFETY: ffi
+        unsafe { CStr::from_ptr(ibv_get_device_name(self.ffi_ptr())) }
+    }
+
+    /// Returns kernel device name
+    #[inline]
+    #[must_use]
+    pub fn name(&self) -> &str {
+        #[allow(clippy::expect_used)]
+        self.c_name().to_str().expect("non-utf8 device name")
```

Comment:
Should we panic? Maybe return option is a better choise. Sometimes users just want to ignore these names.